### DISCUSSION TITLE
Should isValid be true if there are errors?

### DISCUSSION THREAD

**Discussion by Thomas-Smyth** [react-hook-form/react-hook-form#8403] [Q&A]
## Background
I am using the Yup resolver to validate the values of an entire form against a schema at the form level, however, in some circumstances, I have found it easier to validate some of the values in the form with a validation function at the field level. [As it is not possible to use a resolver with the built-in validation](https://react-hook-form.com/api/useform#:~:text=A%20resolver%20cannot%20be%20used%20with%20the%20built%2Din%20validators%20(e.g.%3A%20required%2C%20min%2C%20etc.)), I have had to implement my own resolver to allow me to use the built-in `validate` function validation as a resolver and a utility to combine this resolver and the Yup resolver together. A minimal reproduction of this can be seen [here](https://codesandbox.io/s/magical-snowflake-ny0in7?file=/src/ReactHookForm.js).

I am unsure whether this is correct, however, to my understanding only the fields specified in the `fields`/`names` arguments need to be validated and, if appropriate, have errors returned for by the resolver? I came to this understanding as the errors for unspecified fields were not wiped, which would be expected if the resolver was expected to re-validate the values of every field and re-return all the errors that were still present. Furthermore, as the `validate` function is passed in via the `fields` argument if the fields were not specified it would make it very hard to re-validate them as either the validation function or error message would need to be cached elsewhere.

## Problem
The potential issue I have noticed is that if the resolver only returns errors for the fields requested and there are no errors for those fields then `isValid` is set to true despite there still being errors in the unspecified fields.

To see this in the [CodeSandbox](https://codesandbox.io/s/magical-snowflake-ny0in7?file=/src/ReactHookForm.js), complete the follow steps:
1. Click the first field, type `Hello` into it, and then click off it - this triggers the custom resolver to run with the first field specified and this results in the error from the validation function, which is a result of the field equalling `Hello`, to be returned.
2. Observe that the first field has an error next to it, the `errors` object in the form state contains 1 error, and the `isValid` flag in the form state is `false`.
3. Click the second field, and then click off it - this triggers the custom resolver to run with the second field specified and this results in no errors, as a lack of the first field being specified and the second field having no validation function, being returned.
4. Observe that the first field still has an error next to it, the `errors` object in the form state still contains 1 error, and the `isValid` flag in the form state is now `true` despite there still being errors.
5. Click the first field, and then click off it - this triggers the custom resolver to run with the first field specified again and this results in the error from the validation function to be returned again.
6. Observe that the first field still has an error next to it, the `errors` object in the form state still contains 1 error, and the `isValid` flag in the form state is now `false` again.

Given the `isValid` flag is meant to ["Set to true if the form doesn't have any errors"](https://react-hook-form.com/api/useform/formstate#:~:text=isValid,have%20any%20errors.) it appears to be set incorrectly in steps 3 and 4 when it is set to `true` despite there still being errors in the form.

## Possible Cause
Without diving too deeply into the source of React Hook Form, I hypothesise that `isValid` is being set based on the presence of errors in the response of a resolver called to validate only a specified subset of fields, rather than based on the presence of errors once the errors from the specified subset of fields are merged with the errors from the unspecified subset of fields/the complement to the specified subset of fields. Thus, if there are no errors in a response from a resolver `isValid` is set to true, despite there still being errors in the form relating to the unspecified subset of fields.

## Questions
 * Is my understanding correct that a resolver only needs to evaluate the fields specified?
 * If it is, am I right in saying `isValid` is behaving in an unintended way?
 * If not, how does one wipe errors in their resolver given that returning no error for a field doesn't seem to wipe its error?
 * If not, do you have any suggestions on how to cache validation functions or errors for unspecified fields between runs in the resolver?

**Comment by leapful** [comment]
@Thomas-Smyth 

- Is my understanding correct that a resolver only needs to evaluate the fields specified?
=> No, a resolver will evaluate following the context of its caller, in your example, it's "onBlur" event which is only related to 1 field therefore the only context for this handling is just the specific field that has been on "blur".

- If it is, am I right in saying isValid is behaving in an unintended way?
=>  Not exactly but yes for case "onBlur" which is focusing on just 1 field element that causes missing context. If you submit the whole form, you will see the resolver can get all field elements to validate.

- If not, do you have any suggestions on how to cache validation functions or errors for unspecified fields between runs in the resolver?
=> You can use option "context" of useForm to keep track on errors and update it when needed.

For example: https://codesandbox.io/s/friendly-lucy-jli6uy?file=/src/ReactHookForm.js

**Reply by Thomas-Smyth** [reply]
Thanks for the reply and the example - I've gone for something very similar and it works well.

I am still a little confused about the behaviour of resolvers:
 * If the premise is that all errors in the form, regardless of whether the associated field was specified in the arguments, need to be returned, then surely not returning an error for a field should result in it being removed from the form state/the input?
 * If the premise is that all errors in the form do not need to be returned, providing the associated field was not specified in the arguments, and they are merged with the existing errors, then surely `isValid` should be set based on whether there are errors in the form after the merge?

To me, it seems that part of the logic behaves on one premise and the rest on another?
